### ISSUE TITLE
EWMA support to rolling window functions

### ISSUE THREAD

**Issue by aaelony-aeg** [Rdatatable/data.table#5485]
As mentioned in the following [thread](https://github.com/Rdatatable/data.table/pull/5441#issuecomment-1275062094), this is a new issue to:
 - Represent an enhancement request for providing support for exponential weighting of rolling window averages (and other moments)
 - Enabling EWM mean (and median, sd, other moments) by grouping in `data.table` 

Please see:
 -  This [stackoverflow post](https://stackoverflow.com/questions/65760795/replicate-ewm-pandas-function-in-r) 
 - which mentions this [gist](https://gist.github.com/assuncaolfi/5581528021ac75247a4a1f1c0c3fe12f) .

if I didn't goof converting alpha as an arg to span as an arg, following the [pandas ewm documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ewm.html) the following (or better) might be reasonable:

From the gist above (with minor modification):
```

ewma <- function(x, span) {
  n     <- length(x)
  alpha <- 2 / (span + 1)
  sapply(
    1:n,
    function(i, x, alpha) {
      y <- x[1:i]
      m <- length(y)
      weights <- (1 - alpha)^((m - 1):0)
      ewma <- sum(weights * y) / sum(weights)
    },
    x = x,
    alpha = alpha
  )
}

ewmsd <- function(x, span) {
  n <- length(x)
  alpha <- 2 / (span + 1)
  sapply(
    1:n,
    function(i, x, alpha) {
      y <- x[1:i]
      m <- length(y)
      weights <- (1 - alpha)^((m - 1):0)
      ewma <- sum(weights * y) / sum(weights)
      bias <- sum(weights)^2 / (sum(weights)^2 - sum(weights^2))
      ewmsd <- sqrt(bias * sum(weights * (y - ewma)^2) / sum(weights))
    },
    x = x,
    alpha = alpha
  )
}


dt <- fread("https://raw.githubusercontent.com/jbrownlee/Datasets/master/daily-total-female-births.csv")
names(dt) <- tolower(names(dt))
dt[, emwa_span_9 := ewma(births, span = 9)]
```

```
          date births emwa_span_9
  1: 1959-01-01     35    35.00000
  2: 1959-01-02     32    33.33333
  3: 1959-01-03     30    31.96721
  4: 1959-01-04     31    31.63957
  5: 1959-01-05     44    35.31652
 ---                              
361: 1959-12-27     37    39.88377
362: 1959-12-28     52    42.30702
363: 1959-12-29     48    43.44561
364: 1959-12-30     55    45.75649
365: 1959-12-31     50    46.60519
```

**Comment by jangorecki** [comment]
> Represent an enhancement request for providing support for exponential weighting of rolling window **averages (and other moments)**

Each of them is a big piece of a C code. Considering how feature rich is froll interface now, it adds up to make support for those features in each new rolling function.
If one thinks it is as simple as passing different function `sum`/`mean`/`median` to a rolling function, then no, it's unfortunately not like this. Unless we talk about `frollapply` but this one will never be able to compete to C implemented algos.

From my perspective, and I believe others may be even more conservative. I would focus only on supporting those scenarios in `frollapply`, which means it won't be fast. If one needs speed then I suggest to use R packages that are more specialized on those kids of operations.
I will keep this issue open till we provide `frollapply` solution for it.

> Enabling EWM mean (and median, sd, other moments) by grouping in data.table

If you mean a weight mean (or other functions), not a rolling weighted mean, then all that is already supported in data.table grouping.

**Comment by aaelony-aeg** [comment]
My perspective is that exponential weighting seems to be the next logical step when exploring the impact of rolling window functions.   

I agree with valuing correctness over speed (though speed is nice too).  

It is laborious to verify weighting calculations manually, thus it would be nice to have out-of-the-box EW functions (like [Pandas does](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.ewm.html) or better)

**Comment by jangorecki** [comment]
there are other packages that you can always use to verify, for example TTR: https://cran.r-project.org/package=TTR

**Comment by jangorecki** [comment]
Moreover, EMA is technically not a rolling window calculation, but cumulative window calculation. Due to the fact that the results depends not only on a particular window subset, but on the all past observations. Therefore even `frollapply` will not address EMA properly. Variant of EMA which is a rolling window calculation is described in [this SO question](https://stats.stackexchange.com/questions/44650/a-simpler-way-to-calculate-exponentially-weighted-moving-average), that could be easily implemented if #5295 will be added, but still it is not truly EMA. And considering without having it written in C, it won't scale.
I am leaning towards closing this issue, as it cannot be even nicely addressed.
